src/ImageSharp/Formats/Bmp/BmpDecoderOptions.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+namespace SixLabors.ImageSharp.Formats.Bmp
+{
+    /// <summary>
+    /// Configuration options for decoding Windows Bitmap images.
+    /// </summary>
+    public sealed class BmpDecoderOptions : ISpecializedDecoderOptions
+    {
+        /// <inheritdoc/>
+        public DecoderOptions GeneralOptions { get; set; } = new();
```

Comment:
This is the only specialized decoder option class that actually contains a feature. All the others are dummies, their only purpose to satisfy the pattern introduced. I believe this is alarming.

Reply:
I wanted something that was guaranteed to handle some of the more exotic codes that we'll end up implementing so that we don't have to further breaks. I can imagine both AVIF and Jpeg-XL requiring customization. 

I'm actually very surprised we don't already have something for Tiff and I believe that is most likely an oversight thus far.

Reply:
Even with Bmp + Tiff + AVIF we would have too much of boilerplate of generics and unnecessary empty classes IMO. And Tiff + AVIF are purely hypothetical at the moment, we don't have a proof this feature really needed & will be in fact implemented in the near future. I would prefer a design, where we could YAGNI this, and postpone creating specialized types until the moment they are proven to be actually needed for a format.

On top of that, I realized now, that we are creating an API disparity between decoders and encoders:
Decoders are becoming stateless objects with this PR with separate types for decoding options, while encoder objects carry the encoding options.


I know that I was an advocate of stateless decoders/encoders, but now I disagree with my past self and it feels like the simplest thing would be to move the format specific options back to the decoder objects, so users could do specialized decoding just:

```C#
var decoder = new BmpDecoder { RleSkippedPixelHandling = RleSkippedPixelHandling.Transparent };
using Image img = decoder.Decode(...);
```


Alternatively, we can also consider turning the current abstaction into a pattern, but this would not solve the decoder-encoder API disparity:

```C#
public class ImageDecoder : IImageDecoder {
    public Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken);
}

public class BmpDecoderOptions{
    public DecoderOptions GeneralOptions { get; set; }
}

public BmpDecoder : ImageDecoder {
    public Image<TPixel> DecodeSpecialized(BmpDecoderOptions options, Stream stream, CancellationToken cancellationToken);
}

```

Reply:
> On top of that, I realized now, that we are creating an API disparity between decoders and encoders:
Decoders are becoming stateless objects with this PR with separate types for decoding options, while encoder objects carry the encoding options.

I think we'll end up doing something similar to encoders anyway. 

How would you enforce the options pattern for specialized decoders? I really want to ensure that for specialization we are forced into following something. 

The problem I see with the pattern above is that now you actually have more boilerplate for non-specialized decoding since you now have to implement the resize functionality for each concrete `Decode` implementation.

Reply:
Maybe we can revert the specialized options and make the decoder stateful again?

`DecodeSpecialized` is simply the concrete implementation and each decoder then becomes responsible for passing addition options to the core.

Reply:
> you now have to implement the resize functionality for each concrete `Decode` implementation. 

We have to do it anyways, having a streaming + resizing, low-memory decoding path would be preferable introduce for each codec in 3.0. 

> I can't implement a single abstract way to ensure the decoder honors the resize operation without reimplementing each time.

We can move the current unoptimized implementation to some internal helper method to avoid copypasta.

> Maybe we can revert the specialized options and make the decoder stateful again?

That's basically the first suggestion in my comment. I wonder if we have the same understanding though?

Reply:
I think I understand where you are going. 

This is where I'm currently sitting with an implementation that has custom values and resizes during specialization. 

My only concern here is the separation of options into two separate types. It feels a little clunky 

```c#
/// <summary>
/// Image decoder for generating an image out of a Windows bitmap stream.
/// </summary>
public class BmpDecoder : ImageDecoder
{
    /// <summary>
    /// Gets or sets the value indicating how to deal with skipped pixels,
    /// which can occur during decoding run length encoded bitmaps.
    /// </summary>
    public RleSkippedPixelHandling RleSkippedPixelHandling { get; set; }

    /// <inheritdoc/>
    public override IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Guard.NotNull(stream, nameof(stream));

        return new BmpDecoderCore(this, options).Identify(options.Configuration, stream, cancellationToken);
    }

    /// <inheritdoc/>
    public override Image<TPixel> DecodeSpecialized<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Guard.NotNull(stream, nameof(stream));

        BmpDecoderCore decoder = new(this, options);
        Image<TPixel> image = decoder.Decode<TPixel>(options.Configuration, stream, cancellationToken);

        Resize(options, image);

        return image;
    }

    /// <inheritdoc/>
    public override Image DecodeSpecialized(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
        => this.DecodeSpecialized<Rgba32>(options, stream, cancellationToken);
}
```

Reply:
I'm gonna try your pattern above.

Reply:
OK.... So here's a play on your pattern...

I think this actually solves all our problems. 

- Uses abstraction.
- Add specialized implementation only where it is needed.
- Avoids `Activator`
- Groups options types.

Note. I'm not enforcing a generic type for `ImageDecoderSpecialized<T>` so that we can pass `DecoderOptions` as the generic parameter in the `JpegDecoder` implementation since it requires no additional properties yet requires specialist behavior (_no additional resize_).

```c#
/// <summary>
/// Image decoder for generating an image out of a Windows bitmap stream.
/// </summary>
public class BmpDecoder : ImageDecoderSpecialized<BmpDecoderOptions>
{
    /// <inheritdoc/>
    public override IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Guard.NotNull(options, nameof(options));
        Guard.NotNull(stream, nameof(stream));

        return new BmpDecoderCore(new() { GeneralOptions = options }).Identify(options.Configuration, stream, cancellationToken);
    }

    /// <inheritdoc/>
    public override Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
        => this.DecodeSpecialized<TPixel>(new() { GeneralOptions = options }, stream, cancellationToken);

    /// <inheritdoc/>
    public override Image Decode(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
        => this.DecodeSpecialized(new() { GeneralOptions = options }, stream, cancellationToken);

    /// <inheritdoc/>
    public override Image<TPixel> DecodeSpecialized<TPixel>(BmpDecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Guard.NotNull(options, nameof(options));
        Guard.NotNull(stream, nameof(stream));

        Image<TPixel> image = new BmpDecoderCore(options).Decode<TPixel>(options.GeneralOptions.Configuration, stream, cancellationToken);

        Resize(options.GeneralOptions, image);

        return image;
    }

    /// <inheritdoc/>
    public override Image DecodeSpecialized(BmpDecoderOptions options, Stream stream, CancellationToken cancellationToken)
        => this.DecodeSpecialized<Rgba32>(options, stream, cancellationToken);
```

```c#
/// <summary>
/// The base class for all specialized image decoders.
/// </summary>
/// <typeparam name="T">The type of specialized options.</typeparam>
public abstract class ImageDecoderSpecialized<T> : ImageDecoder
{
    /// <summary>
    /// Decodes the image from the specified stream to an <see cref="Image{TPixel}"/> of a specific pixel type.
    /// </summary>
    /// <typeparam name="TPixel">The pixel format.</typeparam>
    /// <param name="options">The specialized decoder options.</param>
    /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
    /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
    /// <returns>The <see cref="Image{TPixel}"/>.</returns>
    /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
    public abstract Image<TPixel> DecodeSpecialized<TPixel>(T options, Stream stream, CancellationToken cancellationToken)
        where TPixel : unmanaged, IPixel<TPixel>;

    /// <summary>
    /// Decodes the image from the specified stream to an <see cref="Image"/> of a specific pixel type.
    /// </summary>
    /// <param name="options">The specialized decoder options.</param>
    /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
    /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
    /// <returns>The <see cref="Image{TPixel}"/>.</returns>
    /// <exception cref="ImageFormatException">Thrown if the encoded image contains errors.</exception>
    public abstract Image DecodeSpecialized(T options, Stream stream, CancellationToken cancellationToken);
}
```

```c#
/// <summary>
/// The base class for all image decoders.
/// </summary>
public abstract class ImageDecoder : IImageInfoDetector, IImageDecoder
{
    /// <inheritdoc/>
    public abstract IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken);

    /// <inheritdoc/>
    public abstract Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
        where TPixel : unmanaged, IPixel<TPixel>;

    /// <inheritdoc/>
    public abstract Image Decode(DecoderOptions options, Stream stream, CancellationToken cancellationToken);

    /// <summary>
    /// Performs a resize operation against the decoded image. If the target size is not set, or the image size
    /// already matches the target size, the image is untouched.
    /// </summary>
    /// <param name="options">The decoder options.</param>
    /// <param name="image">The decoded image.</param>
    protected static void Resize(DecoderOptions options, Image image)
    {
        if (ShouldResize(options, image))
        {
            ResizeOptions resizeOptions = new()
            {
                Size = options.TargetSize.Value,
                Sampler = KnownResamplers.Box,
                Mode = ResizeMode.Max
            };

            image.Mutate(x => x.Resize(resizeOptions));
        }
    }

    /// <summary>
    /// Determines whether the decoded image should be resized.
    /// </summary>
    /// <param name="options">The decoder options.</param>
    /// <param name="image">The decoded image.</param>
    /// <returns><see langword="true"/> if the image should be resized, otherwise; <see langword="false"/>.</returns>
    private static bool ShouldResize(DecoderOptions options, Image image)
    {
        if (options.TargetSize is null)
        {
            return false;
        }

        Size targetSize = options.TargetSize.Value;
        Size currentSize = image.Size();
        return currentSize.Width != targetSize.Width && currentSize.Height != targetSize.Height;
    }
}
```

Reply:
`BmpDecoder` is now stateless but I think I'm ok with that.

Reply:
After a very quick look: 
- I don't see the need for `ImageDecoderSpecialized<T>` base type. Am I missing something?
- I would actually rename `DecodeSpecialized(BmpDecoderOptions ...)` to `Decode(BmpDecoderOptions ...)`. The argument in the overload makes it obvious it's a codec-specific implementation, I don't feel the need for the verbose name.

> `BmpDecoder` is now stateless but I think I'm ok with that.

If we go with that, we should explore how would a similar separation look like in encoders, and how would we feel about it. (Stateless encoder + option object.)

Reply:
Actually, it can be useful as a potential solution to the problem described in https://github.com/SixLabors/ImageSharp/pull/2180#issuecomment-1203297049.

Reply:
>I don't see why is the ImageDecoderSpecialized<T> base type needed. Am I missing something?

Yeah, I guess that could be simply an interface unless we implement overloads. The `ImageDecoder` actually handles the meat now. 

>I don't feel the need for the verbose name.

The reason it's verbose is so I can do this with the `JpegDecoder`. We need a method that allows IDCT resize only and creating a `JpegDecoderOptions` type as you've identified is not great. 

```c#
/// <summary>
/// Decoder for generating an image out of a jpeg encoded stream.
/// </summary>
public sealed class JpegDecoder : ImageDecoderSpecialized<DecoderOptions>
{
    /// <inheritdoc/>
    public override IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Guard.NotNull(stream, nameof(stream));

        using JpegDecoderCore decoder = new(options);
        return decoder.Identify(options.Configuration, stream, cancellationToken);
    }

    /// <inheritdoc/>
    public override Image<TPixel> Decode<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Image<TPixel> image = this.DecodeSpecialized<TPixel>(options, stream, cancellationToken);

        Resize(options, image);

        return image;
    }

    /// <inheritdoc/>
    public override Image Decode(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        Image image = this.DecodeSpecialized(options, stream, cancellationToken);

        Resize(options, image);

        return image;
    }

    /// <inheritdoc/>
    /// <remarks>
    /// Unlike <see cref="IImageDecoder.Decode{TPixel}(DecoderOptions, Stream, CancellationToken)"/>, when
    /// <see cref="DecoderOptions.TargetSize"/> is passed, the codec may not be able to scale efficiently to
    /// the exact scale factor requested, so returns a size that approximates that scale.
    /// Upscaling is not supported, so the original size will be returned.
    /// </remarks>
    public override Image<TPixel> DecodeSpecialized<TPixel>(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
    {
        using JpegDecoderCore decoder = new(options);
        return decoder.Decode<TPixel>(options.Configuration, stream, cancellationToken);
    }

    /// <inheritdoc/>
    /// <remarks>
    /// Unlike <see cref="IImageDecoder.Decode(DecoderOptions, Stream, CancellationToken)"/>, when
    /// <see cref="DecoderOptions.TargetSize"/> is passed, the codec may not be able to scale efficiently to
    /// the exact scale factor requested, so returns a size that approximates that scale.
    /// Upscaling is not supported, so the original size will be returned.
    /// </remarks>
    public override Image DecodeSpecialized(DecoderOptions options, Stream stream, CancellationToken cancellationToken)
        => this.DecodeSpecialized<Rgb24>(options, stream, cancellationToken);
}
```

Reply:
>  We need a method that allows IDCT resize only 
> ` JpegDecoder : ImageDecoderSpecialized<DecoderOptions>`

To me it looks very unfortunate to use the general `DecoderOptions` as specialized options just to tell that a user wants an `IDCT` resize only. If I would look for such API the last thing I would expect that I just simply need to call a method called `DecodeSpcialized`.

I would rather introduce `JpegDecoderOptions` with a `bool IdctResize` property, or simply keep the "IDCT resize only" feature out of the public API. It's a niche use-case users would only use if they are unhappy with our default `TargetSize` decoding and need better control on the Resize. (Which IMHO would mean that we are doing something wrong and maybe we should reconsider exposing `ResizeOptions`.)

Reply:
If we want maximum flexibility, I would consider something like this for Jpeg:

```C#
public enum JpegResizeMode
{
    Combined, // This is what we do today
    IdctOnly, // IDCT-only to nearest block scale
    ScaleOnly // Opt-out the IDCT part and only Resize. Can be useful in case of quality concerns.
}

public class JpegDecoderOptions
{
    public DecoderOptions GeneralOptions { get; set; }
    public JpegResizeMode ResizeMode { get; set; } = JpegResizeMode.Combined;
}
```

Reply:
OK. I love this!